### ISSUE TITLE
how to use tfrecord data(dict format) (tf.data.dataset) in distribute training?

### ISSUE THREAD

**Issue by wannianlou** [tensorflow/tensorflow#57175]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

source

### Tensorflow Version

tf2.2.0

### Custom Code

Yes

### Python version

3.7

### Current Behaviour?

```shell
I tried to use both mirror and multimirror in distributed training, and i found the input data (from all doucements like tf official tutorial,github and so on )to the strategy.run() must be like ((None,256,256,64)(None)), while the batched data from the fnction below "def make_dataset_for_distribute_train(self, all_paths):" is a dict.  strategy.run() cannot accept dict data for distributed training , so i had to transform it(the output of "make_dataset_for_distribute_train" function) before the distributed training using the
function "def parse_dict2_turple(self,data_tensor" below.
It worked, but the cpu useage is above 300%(before is 200%）and gpu_utils_average is just 10%(before is 75% one gpu). As a result, the distributed training with 4 gpus even slower than normal training. 
Can someone tell me why and how to fix it? really appreciate for your help!!!
```

### Standalone code to reproduce the issue

```shell
def make_dataset_for_distribute_train(self, all_paths):
   dataset = tf.data.TFRecordDataset(all_paths)
   dataset = dataset.map(self.parser, num_parallel_calls=tf.data.experimental.AUTOTUNE)
   dataset = dataset.map(self.make_feed_dict,num_parallel_calls=tf.data.experimental.AUTOTUNE)
   padding_dict = {"a": [20], "b": [None], "c": [30],"d": [13]}

   dataset_for_train = dataset.padded_batch(self.batch_size, padded_shapes=padding_dict,
                                                 drop_remainder=True).prefetch(
            buffer_size=tf.data.experimental.AUTOTUNE)
   return dataset_for_train
    



def parse_dict2_turple(self,data_tensor):

        d0 = data_tensor['a']
        d1 = data_tensor['b']
        d2 = data_tensor['c']
        d3 = data_tensor['d']
  
        return tf.data.Dataset.from_tensor_slices((d0, d1, d2, d3))

        def distributed_train_step(inputs):

            @tf.function
            def distributed_train_step_wrapped(ds):
                loss=0.0
                for k in ds:
                    # tf.print(k)
                    per_replica_average_loss = strategy.run(train_step, args=(k,))
                    loss += strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_average_loss, axis=None)

                return loss
            return distributed_train_step_wrapped(inputs)
            for i in date:
                all_paths = self.parse_input_path(i, epoch)
                logger.info(all_paths)
                dataset = self.make_dataset_for_distribute_train(all_paths)

                total_loss = 0.0
                num_train_batches = 0
                for j in dataset:

                    dataset_parsed = self.parse_dict2_turple(j)
                    dataset_parsed = dataset_parsed.with_options(options)
                    train_dataset_distribute = strategy.experimental_distribute_dataset(dataset_parsed)
                    batch_loss = distributed_train_step(train_dataset_distribute)
                    num_train_batches += 1
                    total_loss += batch_loss
```

### Relevant log output

_No response_</details>

**Comment by wannianlou** [comment]
def make_dataset_for_distribute_train(self, all_paths):
   dataset = tf.data.TFRecordDataset(all_paths)
   dataset = dataset.map(self.parser, num_parallel_calls=tf.data.experimental.AUTOTUNE)
   dataset = dataset.map(self.make_feed_dict,num_parallel_calls=tf.data.experimental.AUTOTUNE)
   padding_dict = {"a": [20], "b": [None], "c": [30],"d": [13]}

   dataset_for_train = dataset.padded_batch(self.batch_size, padded_shapes=padding_dict,
                                                 drop_remainder=True).prefetch(
            buffer_size=tf.data.experimental.AUTOTUNE)
   return dataset_for_train
    



def parse_dict2_turple(self,data_tensor):

        d0 = data_tensor['a']
        d1 = data_tensor['b']
        d2 = data_tensor['c']
        d3 = data_tensor['d']
  
        return tf.data.Dataset.from_tensor_slices((d0, d1, d2, d3))

def distributed_train_step(inputs):

    @tf.function
    def distributed_train_step_wrapped(ds):
        loss=0.0
        for k in ds:
            # tf.print(k)
            per_replica_average_loss = strategy.run(train_step, args=(k,))
            loss += strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_average_loss, axis=None)

        return loss
    return distributed_train_step_wrapped(inputs)
    for i in date:
        all_paths = self.parse_input_path(i, epoch)
        logger.info(all_paths)
        dataset = self.make_dataset_for_distribute_train(all_paths)

        total_loss = 0.0
        num_train_batches = 0
        for j in dataset:

            dataset_parsed = self.parse_dict2_turple(j)
            dataset_parsed = dataset_parsed.with_options(options)
            train_dataset_distribute = strategy.experimental_distribute_dataset(dataset_parsed)
            batch_loss = distributed_train_step(train_dataset_distribute)
            num_train_batches += 1
            total_loss += batch_loss


Aligned code is like this

**Comment by tilakrayal** [comment]
@wannianlou,
To reproduce the issue reported here, could you please provide the complete code and the dataset you are using. 
Also TensorFlow **v2.2** is not actively supported. Could you please update TensorFlow to the latest stable version **_v2.9_** and check if you are facing the same issue. Thank you!
